### PR TITLE
*: merge the runtime stats in time to avoid using too many memory

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/39413

Now tidb need to record some runtime stats for each executor/cop_task and so on, and merge them after the sql finfished.
If there are many cop tasks, the memory can't gc and hold util the sql finfished. Maybe it cause OOM.

Problem Summary:

When the stats will not updated, merge it to the summary stats.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: distsql/select_result.go [hunk 1]
```diff
@@ -456,6 +454,9 @@ func (r *selectResult) Close() error {
 	if respSize > 0 {
 		r.memConsume(-respSize)
 	}
+	if r.stats != nil {
+		defer r.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.RegisterStats(r.rootPlanID, r.stats)
```
Comment (crazycs520) [inline distsql/select_result.go:458]: why use `defer` here?

File: executor/distsql.go [hunk 2]
```diff
@@ -718,6 +718,9 @@ func (e *IndexLookUpExecutor) buildTableReader(ctx context.Context, task *lookup
 
 // Close implements Exec Close interface.
 func (e *IndexLookUpExecutor) Close() error {
+	if e.stats != nil {
+		defer e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.RegisterStats(e.id, e.stats)
```
Comment (crazycs520) [inline executor/distsql.go:722]: why use defer?

File: distsql/select_result.go (hunk 1 above)
Comment (wshwsh12) [reply distsql/select_result.go:458]: Use defer to ensure that the `Close()`(make sure all goroutines have been exit in some parallel executer) is finished before registering stats.

**Review by crazycs520 — APPROVED** [review]

**Review by XuHuaiyu — APPROVED** [review]

**Comment by hawkingrei** [comment]
@wshwsh12 
```
    execdetails_test.go:228: 
        	Error Trace:	/home/jenkins/.tidb/tmp/63a9840cd0739f2c243bb46478607469/sandbox/linux-sandbox/3543/execroot/__main__/bazel-out/k8-fastbuild/bin/util/execdetails/execdetails_test_/execdetails_test.runfiles/__main__/util/execdetails/execdetails_test.go:228
        	Error:      	Not equal: 
        	            	expected: "tiflash_task:{proc max:2ns, min:1ns, avg: 1ns, p80:2ns, p95:2ns, iters:3, tasks:2, threads:2}, tiflash_scan:{dmfile:{total_scanned_packs:1, total_skipped_packs:0, total_scanned_rows:8192, total_skipped_rows:0, total_rough_set_index_load_time: 15ms, total_read_time: 202ms}, total_create_snapshot_time: 40ms}"
        	            	actual  : "tiflash_task:{proc max:2ns, min:1ns, avg: 1ns, p80:2ns, p95:2ns, iters:3, tasks:2, threads:2}"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-tiflash_task:{proc max:2ns, min:1ns, avg: 1ns, p80:2ns, p95:2ns, iters:3, tasks:2, threads:2}, tiflash_scan:{dmfile:{total_scanned_packs:1, total_skipped_packs:0, total_scanned_rows:8192, total_skipped_rows:0, total_rough_set_index_load_time: 15ms, total_read_time: 202ms}, total_create_snapshot_time: 40ms}
        	            	+tiflash_task:{proc max:2ns, min:1ns, avg: 1ns, p80:2ns, p95:2ns, iters:3, tasks:2, threads:2}
        	Test:       	TestCopRuntimeStatsForTiFlash
```

**Pull request merged**